### PR TITLE
Improve enum support with EnumField and EnumFilter

### PR DESCRIPTION
**Pull request by acrobat**
I've added a EnumField type to render database enum values correctly as an Enum can't be converted to string (for the string type). In this field I've used the same logic as in the Symfony Enum form type, eg. if it's an translatable enum, translate it first otherwise just render the ->value of the enum

The same goes for the filter where we use the Symfony Enum form type to render the filter, the only config it needs is the enum class used to render the options.

### CODE REVIEW

File: src/Component/spec/FieldTypes/EnumFieldTypeSpec.php [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+use Sylius\Component\Grid\DataExtractor\DataExtractorInterface;
+use Sylius\Component\Grid\Definition\Field;
+use Sylius\Component\Grid\FieldTypes\FieldTypeInterface;
+use Sylius\Component\Grid\Tests\Dummy\TestEnum;
+use Sylius\Component\Grid\Tests\Dummy\TestNonBackedEnum;
+use Sylius\Component\Grid\Tests\Dummy\TestTranslatableEnum;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+use Symfony\Contracts\Translation\TranslatorInterface;
+
+final class EnumFieldTypeSpec extends ObjectBehavior
```
Comment (loic425) [inline src/Component/spec/FieldTypes/EnumFieldTypeSpec.php]: Please, create a PHPUnit test instead.
We are currently removing all Phpspec tests.

File: composer.json [hunk 2]
```diff
@@ -40,6 +40,7 @@
         "symfony/http-kernel": "^6.4 || ^7.0",
         "symfony/options-resolver": "^6.4 || ^7.0",
         "symfony/property-access": "^6.4 || ^7.0",
+        "symfony/translation": "^6.4 || ^7.0",
```
Comment (loic425) [inline composer.json]: please move on dev requirements to keep it optional.

File: src/Bundle/Resources/config/services/field_types.xml [hunk 3]
```diff
@@ -34,5 +34,12 @@
             <tag name="sylius.grid_field" type="string" />
         </service>
         <service id="Sylius\Component\Grid\FieldTypes\StringFieldType" alias="sylius.grid_field.string" />
+
+        <service id="sylius.grid_field.enum" class="Sylius\Component\Grid\FieldTypes\EnumFieldType">
+            <argument type="service" id="sylius.grid.data_extractor" />
+            <argument type="service" id="translator" />
```
Comment (loic425) [inline src/Bundle/Resources/config/services/field_types.xml]:
```suggestion
            <argument type="service" id="translator" on-invalid="null" />
```

File: src/Component/FieldTypes/EnumFieldType.php [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+final class EnumFieldType implements FieldTypeInterface
+{
+    private DataExtractorInterface $dataExtractor;
+    private TranslatorInterface $translator;
+
+    public function __construct(DataExtractorInterface $dataExtractor, TranslatorInterface $translator)
+    {
+        $this->dataExtractor = $dataExtractor;
+        $this->translator = $translator;
+    }
```
Comment (loic425) [inline src/Component/FieldTypes/EnumFieldType.php]:
```suggestion
    public function __construct(
        private DataExtractorInterface $dataExtractor, 
        private ?TranslatorInterface $translator = null,
   ) {
    }
```

File: src/Component/FieldTypes/EnumFieldType.php [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+        if (!$enum instanceof \BackedEnum) {
+            return $enum->name;
+        }
+
+        if (!$enum instanceof TranslatableInterface) {
+            return (string) $enum->value;
+        }
+
+        return $enum->trans($this->translator);
```
Comment (loic425) [inline src/Component/FieldTypes/EnumFieldType.php]:
```suggestion
if (null === $this->trans) {
            throw new LogicException('You have configured a translatable enum, but Symfony translator is not available. Try running "composer require symfony/translator".');
        }

        return $enum->trans($this->translator);
```

File: src/Component/FieldTypes/EnumFieldType.php [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Component\Grid\FieldTypes;
+
+use Sylius\Component\Grid\DataExtractor\DataExtractorInterface;
+use Sylius\Component\Grid\Definition\Field;
```
Comment (loic425) [inline src/Component/FieldTypes/EnumFieldType.php:17]:
```suggestion
use Sylius\Component\Grid\Definition\Field;
use Sylius\Component\Exception\LogicException;
```

File: src/Component/Filter/EnumFilter.php [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+declare(strict_types=1);
+
+namespace Sylius\Component\Grid\Filter;
+
+use Sylius\Component\Grid\Data\DataSourceInterface;
+use Sylius\Component\Grid\Filtering\FilterInterface;
+
+final class EnumFilter implements FilterInterface
+{
+    public function apply(DataSourceInterface $dataSource, string $name, $data, array $options): void
```
Comment (loic425) [inline src/Component/Filter/EnumFilter.php]: this code is the same on as the SelectFilter, so I think we could use the SelectFilter instead.

**Review by loic425 — CHANGES_REQUESTED** [review]

**Comment by acrobat** [comment]
Thanks for the review @loic425! I've processed your feedback and converted the spec test to a phpunit test!

File: src/Component/FieldTypes/EnumFieldType.php [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        if (!$enum instanceof \BackedEnum) {
+            return $enum->name;
+        }
+
+        if (!$enum instanceof TranslatableInterface) {
+            return (string) $enum->value;
+        }
+
+        if (null === $this->translator) {
+            throw new \LogicException('You have configured a translatable enum, but Symfony translator is not available. Try running "composer require symfony/translator".');
```
Comment (loic425) [inline src/Component/FieldTypes/EnumFieldType.php]: Please use LogicException from the Grid namespace.

File: src/Component/FieldTypes/EnumFieldType.php (hunk 6 above)
Comment (loic425) [reply src/Component/FieldTypes/EnumFieldType.php:17]: You need to use this exception, not the php native one.

**Review by loic425 — CHANGES_REQUESTED** [review]

File: src/Component/FieldTypes/EnumFieldType.php (hunk 8 above)
Comment (acrobat) [reply src/Component/FieldTypes/EnumFieldType.php]: Oops, I missed that one. Fixed!

**Comment by loic425** [comment]
A few things to add:
- ~~A test of the admin user grids on GridUiTest~~
- admin user grid configuration on config (both yaml and php).
    - yaml one https://github.com/Sylius/SyliusGridBundle/blob/1.14/tests/Application/config/sylius/grids.yaml
    - php one https://github.com/Sylius/SyliusGridBundle/tree/1.14/tests/Application/config/sylius/grids

File: tests/Application/config/sylius/grids.yaml [hunk 9]
```diff
@@ -208,3 +208,30 @@ sylius_grid:
...
+                status:
+                    type: enum
+                    options:
+                        enum_class: App\Enum\AdminUserStatusEnum
+                        multiple: true
+            fields:
+                title:
+                    type: callable
+                    options:
+                        callable: "callable:strtoupper"
```
Comment (loic425) [inline tests/Application/config/sylius/grids.yaml]:
```suggestion
                    type: string
```

**Review by loic425 — CHANGES_REQUESTED** [review]

File: src/Component/FieldTypes/EnumFieldType.php [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    public function render(Field $field, $data, array $options): string
+    {
+        $enum = $this->dataExtractor->get($field, $data);
+
+        if (!$enum instanceof \BackedEnum) {
+            return $enum->name;
+        }
+
+        if (!$enum instanceof TranslatableInterface) {
+            return (string) $enum->value;
```
Comment (loic425) [inline src/Component/FieldTypes/EnumFieldType.php]: I think we'll need a try catch here.

Like this one implemented by Florian on Callable (you'll need adapt the code, of course).
```
try {
    $value = (string) $value;
} catch (\Throwable $e) {
    throw new UnexpectedValueException(\sprintf(
        'Callable field (name "%s") returned value could not be converted to string: "%s".',
        $field->getName(),
        $e->getMessage(),
    ));
}
```

Comment (acrobat) [reply src/Component/FieldTypes/EnumFieldType.php]: Can't we safely assume a cast will work as the Backenum::$value property only allows int or string

> A Backed Enum may be backed by types of int or string, and a given enumeration supports only a single type at a time (that is, no union of int|string). 
https://www.php.net/manual/en/language.enumerations.backed.php

File: src/Component/FieldTypes/EnumFieldType.php [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+        private ?TranslatorInterface $translator = null,
+    ) {
+    }
+
+    public function render(Field $field, $data, array $options): string
+    {
+        $enum = $this->dataExtractor->get($field, $data);
+
+        if (!$enum instanceof \UnitEnum) {
+            return '';
```
Comment (loic425) [inline src/Component/FieldTypes/EnumFieldType.php:37]: Should we throw an exception instead?

File: src/Component/Tests/Unit/FieldTypes/EnumFieldTypeTest.php [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+
+    public function testAGridFieldType(): void
+    {
+        $this->assertInstanceOf(FieldTypeInterface::class, $this->enumFieldType);
+    }
+
+    public function testUsesDataExtractorToObtainDataAndRendersIt(): void
+    {
+        /** @var Field|MockObject $fieldMock */
+        $fieldMock = $this->createMock(Field::class);
```
Comment (loic425) [inline src/Component/Tests/Unit/FieldTypes/EnumFieldTypeTest.php]: I think you can use the real instance instead.
EnumField::create...

Comment (acrobat) [reply src/Component/Tests/Unit/FieldTypes/EnumFieldTypeTest.php]: Sure, I now based it on the other unit tests from https://github.com/Sylius/SyliusGridBundle/pull/399/files#diff-57fa8d133cf70faa3b036d9b7d5a703885049c50865b1b6db255fd43b25031ddR47

But I've updated the tests to reflect this!

File: src/Component/FieldTypes/EnumFieldType.php (hunk 11 above)
Comment (acrobat) [reply src/Component/FieldTypes/EnumFieldType.php:37]: I added this case for null values but I solved it in another way (similar to the datetime field type) and added tests for the extra cases (null and invalid values)

**Comment by acrobat** [comment]
@loic425 I've addressed your latest comments!

File: src/Component/FieldTypes/EnumFieldType.php [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+        if (!$enum instanceof TranslatableInterface) {
+            return (string) $enum->value;
+        }
+
+        if (null === $this->translator) {
+            throw new LogicException('You have configured a translatable enum, but Symfony translator is not available. Try running "composer require symfony/translator".');
+        }
+
+        return $enum->trans($this->translator);
```
Comment (Prometee) [inline src/Component/FieldTypes/EnumFieldType.php]: Maybe I'm wrong since I'm not using the Enum so often, but does a non backed enum can have `TranslatableInterface`, if yes then this service must check first for the `TranslatableInterface` then fallback to the `->name` or `->value`.

WDYT?

Comment (loic425) [reply src/Component/FieldTypes/EnumFieldType.php]: Chat GPT warns about that.

image

Comment (loic425) [reply src/Component/FieldTypes/EnumFieldType.php]: his suggestion
image

Comment (acrobat) [reply src/Component/FieldTypes/EnumFieldType.php]: It is indeed possible that a non backed enum is translatable, but I think in the cases of grids which fetches the data from the DB it will always be a backed enum. But to be fully feature complete I've switched the checks around to make that possible!

Comment (Prometee) [reply src/Component/FieldTypes/EnumFieldType.php]: @acrobat Don't forget Grid can be used with simple DTO out of any DB 😉

**Review by loic425 — APPROVED** [review]

**Review by Prometee — APPROVED** [review]

File: src/Component/FieldTypes/EnumFieldType.php (hunk 13 above)
Comment (acrobat) [reply src/Component/FieldTypes/EnumFieldType.php]: Yeah true, that's now supported! 🎉

**Review by NoResponseMate — APPROVED** [review]
